File: examples/demos/scenes/shapes.cpp
/*
 * Copyright (c) 2011, The DART development contributors
 * All rights reserved.
 *
 * The list of contributors can be found at:
 *   https://github.com/dartsim/dart/blob/main/LICENSE
 *
 * This file is provided under the "BSD-style" License.
 */

#include "scenes.hpp"

#include <dart/gui/panel.hpp>
#include <dart/gui/viewer.hpp>

#include <dart/simulation/world.hpp>

#include <dart/dynamics/body_node.hpp>
#include <dart/dynamics/box_shape.hpp>
#include <dart/dynamics/free_joint.hpp>
#include <dart/dynamics/inertia.hpp>
#include <dart/dynamics/skeleton.hpp>
#include <dart/dynamics/sphere_shape.hpp>
#include <dart/dynamics/weld_joint.hpp>

#include <Eigen/Geometry>

#include <memory>
#include <string>

namespace dart::examples::demos {

namespace {

dart::dynamics::SkeletonPtr createGround()
{
  auto ground = dart::dynamics::Skeleton::create("ground");
  auto jointAndBody
      = ground->createJointAndBodyNodePair<dart::dynamics::WeldJoint>();
  auto* joint = jointAndBody.first;
  auto* body = jointAndBody.second;

  Eigen::Isometry3d transform = Eigen::Isometry3d::Identity();
  transform.translation() = Eigen::Vector3d(0.0, 0.0, -0.05);
  joint->setTransformFromParentBodyNode(transform);

  const auto shape = std::make_shared<dart::dynamics::BoxShape>(
      Eigen::Vector3d(10.0, 10.0, 0.1));
  auto* shapeNode = body->createShapeNodeWith<
      dart::dynamics::VisualAspect,
      dart::dynamics::CollisionAspect,
      dart::dynamics::DynamicsAspect>(shape);
  shapeNode->getVisualAspect()->setColor(Eigen::Vector3d(0.8, 0.8, 0.8));

  return ground;
}

dart::dynamics::SkeletonPtr createBox(
    const std::string& name,
    const Eigen::Vector3d& position,
    const Eigen::Vector3d& color)
{
  auto skeleton = dart::dynamics::Skeleton::create(name);
  auto* body = skeleton->createJointAndBodyNodePair<dart::dynamics::FreeJoint>()
                   .second;

  Eigen::Isometry3d transform = Eigen::Isometry3d::Identity();
  transform.translation() = position;
  body->getParentJoint()->setTransformFromParentBodyNode(transform);

  const auto shape = std::make_shared<dart::dynamics::BoxShape>(
      Eigen::Vector3d::Constant(0.24));
  auto* shapeNode = body->createShapeNodeWith<
      dart::dynamics::VisualAspect,
      dart::dynamics::CollisionAspect,
      dart::dynamics::DynamicsAspect>(shape);
  shapeNode->getVisualAspect()->setColor(color);
  body->setInertia(
      dart::dynamics::Inertia(
          1.0, Eigen::Vector3d::Zero(), shape->computeInertia(1.0)));
  return skeleton;
}

dart::dynamics::SkeletonPtr createSphere(
    const std::string& name,
    const Eigen::Vector3d& position,
    const Eigen::Vector3d& color)
{
  auto skeleton = dart::dynamics::Skeleton::create(name);
  auto* body = skeleton->createJointAndBodyNodePair<dart::dynamics::FreeJoint>()
                   .second;

  Eigen::Isometry3d transform = Eigen::Isometry3d::Identity();
  transform.translation() = position;
  body->getParentJoint()->setTransformFromParentBodyNode(transform);

  const auto shape = std::make_shared<dart::dynamics::SphereShape>(0.14);
  auto* shapeNode = body->createShapeNodeWith<
      dart::dynamics::VisualAspect,
      dart::dynamics::CollisionAspect,
      dart::dynamics::DynamicsAspect>(shape);
  shapeNode->getVisualAspect()->setColor(color);
  body->setInertia(
      dart::dynamics::Inertia(
          1.0, Eigen::Vector3d::Zero(), shape->computeInertia(1.0)));
  return skeleton;
}

} // namespace

dart::gui::ApplicationOptions makeShapesScene()
{
  auto world = dart::simulation::World::create("shapes");
  world->setGravity(Eigen::Vector3d(0.0, 0.0, -9.81));
  world->addSkeleton(createGround());
  world->addSkeleton(createBox("box_a", {-0.5, 0.0, 0.4}, {0.90, 0.45, 0.20}));
  world->addSkeleton(
      createSphere("sphere_a", {0.0, 0.0, 0.5}, {0.20, 0.55, 0.90}));
  world->addSkeleton(createBox("box_b", {0.5, 0.0, 0.6}, {0.40, 0.80, 0.35}));

  dart::gui::ApplicationOptions options;
  options.world = world;

  dart::gui::Panel panel;
  panel.title = "Shapes";
  panel.initialPosition = std::array<double, 2>{312.0, 12.0};
  panel.buildWithContext
      = [](dart::gui::PanelBuilder& builder, dart::gui::PanelContext&) {
          builder.text("Assorted primitive shapes settling on the ground.");
        };
  options.panels.push_back(panel);

  dart::gui::OrbitCamera camera;
  camera.target = Eigen::Vector3d(0.0, 0.0, 0.30);
  camera.distance = 3.2;
  options.camera = camera;
  return options;
}

} // namespace dart::examples::demos

File: python/examples/demos/scenes/shapes.py
"""Shapes scene: a box, a sphere, and another box settling on the ground.

Mirrors examples/demos/scenes/shapes.cpp.
"""

from __future__ import annotations

import numpy as np

import dartpy as dart

from ..runner import PythonDemoScene, SceneSetup


def _make_ground() -> "dart.Skeleton":
    skel = dart.Skeleton("ground")
    joint, body = skel.create_weld_joint_and_body_node_pair()
    transform = np.eye(4)
    transform[:3, 3] = (0.0, 0.0, -0.05)
    joint.set_transform_from_parent_body_node(transform)
    shape = dart.BoxShape(np.array([10.0, 10.0, 0.1]))
    sn = body.create_shape_node(shape)
    sn.create_visual_aspect().set_color([0.8, 0.8, 0.8])
    sn.create_collision_aspect()
    sn.create_dynamics_aspect()
    return skel


def _make_box(
    name: str, position: tuple[float, float, float], color: tuple[float, float, float]
) -> "dart.Skeleton":
    skel = dart.Skeleton(name)
    _, body = skel.create_free_joint_and_body_node_pair()
    tf = np.eye(4)
    tf[:3, 3] = position
    body.get_parent_joint().set_transform_from_parent_body_node(tf)

    size = np.array([0.24, 0.24, 0.24])
    shape = dart.BoxShape(size)
    sn = body.create_shape_node(shape)
    sn.create_visual_aspect().set_color(list(color))
    sn.create_collision_aspect()
    sn.create_dynamics_aspect()
    body.set_inertia(dart.Inertia(
        1.0,
        np.zeros(3),
        dart.BoxShape.compute_inertia_of(size, 1.0),
    ))
    return skel


def _make_sphere(
    name: str, position: tuple[float, float, float], color: tuple[float, float, float]
) -> "dart.Skeleton":
    skel = dart.Skeleton(name)
    _, body = skel.create_free_joint_and_body_node_pair()
    tf = np.eye(4)
    tf[:3, 3] = position
    body.get_parent_joint().set_transform_from_parent_body_node(tf)

    shape = dart.SphereShape(0.14)
    sn = body.create_shape_node(shape)
    sn.create_visual_aspect().set_color(list(color))
    sn.create_collision_aspect()
    sn.create_dynamics_aspect()
    body.set_inertia(dart.Inertia(
        1.0,
        np.zeros(3),
        dart.SphereShape.compute_inertia_of(0.14, 1.0),
    ))
    return skel


def build() -> SceneSetup:
    world = dart.World("shapes")
    world.set_gravity([0.0, 0.0, -9.81])
    world.add_skeleton(_make_ground())
    world.add_skeleton(_make_box("box_a", (-0.5, 0.0, 0.4), (0.90, 0.45, 0.20)))
    world.add_skeleton(_make_sphere("sphere_a", (0.0, 0.0, 0.5), (0.20, 0.55, 0.90)))
    world.add_skeleton(_make_box("box_b", (0.5, 0.0, 0.6), (0.40, 0.80, 0.35)))
    return SceneSetup(world=world, info={})


SCENE = PythonDemoScene(
    id="shapes",
    title="Shapes",
    category="Visualization",
    summary="Assorted primitive shapes resting on the ground.",
    build=build,
)
